pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -886,6 +888,35 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
+    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+        if (!conf.isPerMessageRedeliverUnacknowledged()
+                || messageIds.size() > conf.getMaxRedeliverUnacknowledgedMessagesBatch()) {
+            redeliverUnacknowledgedMessages();
+            return;
+        }
+        ClientCnx cnx = cnx();
+        if (isConnected() && cnx.getRemoteEndpointProtocolVersion() >= ProtocolVersion.v2.getNumber()) {
+            MessageIdData.Builder builder = MessageIdData.newBuilder();
+            List<MessageIdData> messageIdDatas = messageIds.stream()
+                    .map(messageId -> {
+                        builder.setPartition(messageId.getPartitionIndex());
+                        builder.setLedgerId(messageId.getLedgerId());
+                        builder.setEntryId(messageId.getEntryId());
```

Comment:
That's true, in case of batched messages, the application will use ack multiple `BatchMessageImpl`. When all the messages within a single batch are acked, that will trigger the ack to the broker based on the whole batch `MessageIdImpl`